<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Cured - TBRU</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Cured";
    var mkdocs_page_input_path = "Cured.md";
    var mkdocs_page_url = "/Cured/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> TBRU</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">16S data</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../GettingStarted/">Getting Started</a>
                </li>
                <li class="">
                    
    <a class="" href="../Treatment/">Treatment</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">Cured</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#hrze-causes-dysbiosis">HRZE causes dysbiosis</a></li>
    
        <ul>
        
            <li><a class="toctree-l4" href="#jump-right-to-subsetting">Jump right to subsetting...</a></li>
        
            <li><a class="toctree-l4" href="#deseq-normalization">DESeq normalization</a></li>
        
            <li><a class="toctree-l4" href="#deseq-normalization-plots">DESeq normalization plots</a></li>
        
            <li><a class="toctree-l4" href="#run-lefse-for-ltbi-vs-cured">Run LEfSE for LTBI vs Cured</a></li>
        
        </ul>
    

    </ul>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Metagenomics</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../Treatment_metagenomics/">Treatment</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../about/">About</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">TBRU</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>16S data &raquo;</li>
        
      
    
    <li>Cured</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="hrze-causes-dysbiosis">HRZE causes dysbiosis</h1>
<p>This tutorial is a walkthrough of the data analysis from:</p>
<p>Antibiotic treatment for Tuberculosis induces a profound dysbiosis of the microbiome that persists long after therapy is completed</p>
<p>Scientific Reports 7, Article number: 10767 (2017)</p>
<p><a href="https://www.nature.com/articles/s41598-017-10346-6">doi: 10.1038/s41598-017-10346-6</a></p>
<p>It shows how to take microbiome data and reproduce the figures from this paper.</p>
<h2 id="jump-right-to-subsetting">Jump right to subsetting...</h2>
<p>See the previous section (Treatment) for how to generate the Phyloseq object</p>
<pre><code class="r">#IGRA positive vs cured (IGRA+ vs cured)
phy_pos_cured &lt;- subset_samples(phy,!is.na(IGRA) &amp; IGRA!=&quot;negative&quot;)
phy_pos_cured &lt;- subset_samples(phy_pos_cured,!is.na(IGRA) &amp; IGRA!=&quot;&quot;)
phy_pos_cured &lt;- subset_samples(phy_pos_cured,!is.na(TB_status) &amp; TB_status!=&quot;treatment&quot;)
phy_pos_cured &lt;- subset_samples(phy_pos_cured,!is.na(Group6_TB_category) &amp; Group6_TB_category!=&quot;family_contact&quot;)
phy_pos_cured &lt;- subset_samples(phy_pos_cured,!is.na(Group6_TB_category) &amp; Group6_TB_category!=&quot;treatment&quot;)
phy_pos_cured &lt;- subset_samples(phy_pos_cured, age &lt; 30)
</code></pre>

<pre><code class="r">#Variance of OTU abundances
hist(log10(apply(otu_table(phy_pos_cured), 1, var)), xlab = &quot;log10(variance)&quot;, main = &quot;Variance distribution for LTBI/Cured&quot;)
</code></pre>

<p><img alt="image of histogram abundance" src="../img/Cured/Variance_LTBI_Cured.png" /></p>
<h2 id="deseq-normalization">DESeq normalization</h2>
<pre><code class="r">library(&quot;DESeq2&quot;)
phy_DESeq &lt;- phy_pos_cured

t &lt;- get.tax(phy_DESeq) %&gt;% mutate(PhySpec=paste(Species,otu))
taxa_names(phy_DESeq) &lt;- t$PhySpec
taxa_names(phy_DESeq) &lt;- gsub(&quot; &quot;,&quot;_&quot;,taxa_names(phy_DESeq))
taxa_names(phy_DESeq) &lt;- gsub(&quot;=&quot;,&quot;_&quot;,taxa_names(phy_DESeq))
taxa_names(phy_DESeq) &lt;- gsub(&quot;;&quot;,&quot;_&quot;,taxa_names(phy_DESeq))

dig &lt;- phyloseq_to_deseq2(phy_DESeq, ~ TB_status) #replace this with any sample variable(s)

#calculate geometric mean
gm_mean &lt;- function(x, na.rm=TRUE){
exp(sum(log(x[x &gt; 0]), na.rm=na.rm) / length(x))
}
geoMeans &lt;- apply(counts(dig), 1, gm_mean)
dig &lt;- estimateSizeFactors(dig, geoMeans = geoMeans)
dig &lt;- estimateDispersions(dig)
dig &lt;- DESeq(dig,fitType= &quot;local&quot;)

# diagvst &lt;- getVarianceStabilizedData(dig)
# dim(diagvst)
# otu_table(phyIGRA_pos_treatment) &lt;- otu_table(diagvst, taxa_are_rows = TRUE)

res &lt;- results(dig)
#res$pfdr &lt;- p.adjust(res$pvalue,method=&quot;fdr&quot;)
res &lt;- res[order(res$padj, na.last=NA), ]
alpha &lt;- 0.05
sigtab &lt;- res[(res$padj &lt; alpha), ]
sigtab &lt;- cbind(as(sigtab, &quot;data.frame&quot;), as(tax_table(phy_DESeq)[rownames(sigtab), ], &quot;matrix&quot;))
head(sigtab) #view the data

posigtab &lt;- sigtab[sigtab[, &quot;log2FoldChange&quot;] &gt; 1, ]
posigtab &lt;- posigtab[, c(&quot;baseMean&quot;, &quot;log2FoldChange&quot;, &quot;lfcSE&quot;, &quot;padj&quot;, &quot;Phylum&quot;, &quot;Class&quot;, &quot;Family&quot;, &quot;Genus&quot;)]
</code></pre>

<h2 id="deseq-normalization-plots">DESeq normalization plots</h2>
<pre><code class="r">library(&quot;ggplot2&quot;)
theme_set(theme_bw())
sigtabgen &lt;- subset(sigtab, !is.na(Genus))
sigtabgen &lt;- subset(sigtab, !is.na(Phylum))
# Phylum order
x &lt;- tapply(sigtabgen$log2FoldChange, sigtabgen$Phylum, function(x) max(x))
x &lt;- sort(x, TRUE)
sigtabgen$Phylum = factor(as.character(sigtabgen$Phylum), levels=names(x))
# Genus order
x &lt;- tapply(sigtabgen$log2FoldChange, sigtabgen$Genus, function(x) max(x))
x &lt;- sort(x, TRUE)
sigtabgen$Genus &lt;- factor(as.character(sigtabgen$Genus), levels=names(x))

phylumcolors &lt;- c(&quot;hotpink&quot;,&quot;brown&quot;,&quot;lightblue&quot;,&quot;purple&quot;)
ggplot(sigtabgen, aes(x = Genus, y = log2FoldChange, color = Phylum)) + geom_point(size = 6) +
theme(axis.text.x = element_text(angle = -45, hjust = 0, vjust = 1)) +
theme(axis.text=element_text(size=12,face=&quot;bold&quot;),legend.text=element_text(size=12))
</code></pre>

<h2 id="run-lefse-for-ltbi-vs-cured">Run LEfSE for LTBI vs Cured</h2>
<pre><code class="r">phy.lefse &lt;- phy_DESeq
results_folder &lt;- &quot;~/Desktop/uparse_july/&quot;
class &lt;- &quot;TB_status&quot;
subclass&lt;-FALSE
subject&lt;-&quot;sample&quot;
anova.alpha&lt;-0.05 #this is the important p value
wilcoxon.alpha&lt;-0.05
lda.cutoff&lt;-2.5
wilcoxon.within.subclass &lt;- TRUE
one.against.one &lt;- T
mult.test.correction &lt;- 0
make.lefse.plots &lt;- FALSE
by_otus &lt;- FALSE
#
sample.data &lt;- phyloseq::sample_data(phy.lefse) %&gt;% data.frame(stringsAsFactors = FALSE)
sample.data$sample &lt;- rownames(sample.data)
#
keepvars &lt;- c(&quot;sample&quot;,&quot;TB_status&quot;)
keepvars &lt;- unique(keepvars[!is.na(keepvars)])
lefse.samp &lt;- sample.data[, keepvars]
#
sample0 &lt;- t(lefse.samp) %&gt;% as.matrix()
colnames(sample0) &lt;- sample0[1,]
sample0 &lt;- as.data.frame(sample0)
#
data0 &lt;- otu_table(phy.lefse) %&gt;% as.data.frame()
data1 &lt;- data0 %&gt;% as.data.table(keep.rownames=T)
sample1 &lt;- sample0 %&gt;% as.data.table(keep.rownames=T)
common &lt;- intersect(colnames(data1), colnames(sample1))
pre.lefse &lt;- rbind(sample1, data1,fill=T) %&gt;% t() %&gt;% na.omit() %&gt;% t()

#writes table for LEfSe
write.table(pre.lefse,file =paste(results_folder,&quot;lefse.txt&quot;,sep=&quot;&quot;),sep = &quot;\t&quot;,row.names = FALSE,col.names = FALSE,quote = FALSE)
opt.class &lt;- paste(&quot;-c&quot;, which(keepvars %in% class))
opt.subclass &lt;- ifelse(is.na(subclass), &quot;&quot;, paste(&quot;-s&quot;, which(keepvars %in%
subclass)))
opt.subject &lt;- ifelse(is.na(subject), &quot;&quot;, paste(&quot;-u&quot;, which(keepvars %in%
subject)))
format.command &lt;- paste(paste(&quot;format_input.py &quot;,results_folder,&quot;lefse.txt &quot;,results_folder,&quot;lefse.in&quot;,sep=&quot;&quot;),
opt.class, opt.subject, &quot;-o 1000000&quot;)
format.command
#&quot;format_input.py ~/Desktop/uparse_july/lefse.txt ~/Desktop/uparse_july/lefse.in -c 2 -u 1 -o 1000000&quot;
system(format.command)
lefse.command &lt;- paste(paste(&quot;~/miniconda2/bin/python ~/lefse/run_lefse.py &quot;,results_folder,&quot;lefse.in &quot; ,results_folder, &quot;lefse.res&quot;,sep=&quot;&quot;),
&quot;-a&quot;, anova.alpha, &quot;-w&quot;, wilcoxon.alpha, &quot;-l&quot;, lda.cutoff,
&quot;-e&quot;, as.numeric(wilcoxon.within.subclass), &quot;-y&quot;, as.numeric(one.against.one),
&quot;-s&quot;, mult.test.correction)
lefse.command
#&quot;~/miniconda2/bin/python ~/lefse/run_lefse.py ~/Desktop/uparse_july/lefse.in ~/Desktop/uparse_july/lefse.res -a 0.05 -w 0.05 -l 3 -e 1 -y 1 -s 0&quot;
system(lefse.command) #this will print out the number of significant OTUs (make sure it's not 0)
lefse.out &lt;- read.table(paste(results_folder,&quot;lefse.res&quot;,sep=&quot;&quot;), header = FALSE, sep = &quot;\t&quot;)
names(lefse.out)&lt;-c(&quot;taxon&quot;,&quot;log.max.pct&quot;,&quot;direction&quot;,&quot;lda&quot;,&quot;p.value&quot;)
(lefse.out&lt;-na.omit(lefse.out))
Palette_LTBI_cured &lt;- c(&quot;#e7298a&quot;,&quot;#377eb8&quot;)
if(length(unique(lefse.out$direction))&lt;3){
lefse.out$lda[lefse.out$direction==unique(lefse.out$direction)[1]] &lt;-
-1*lefse.out$lda[lefse.out$direction==unique(lefse.out$direction)[1]]
}
lefse.out$taxon&lt;-factor(lefse.out$taxon,levels=lefse.out$taxon[order(lefse.out$lda)])
g1&lt;-ggplot(data=lefse.out,aes(x=taxon,y=lda,color=direction,fill=direction))+
geom_bar(stat=&quot;identity&quot;)+
coord_flip()+
theme_base()
if(length(unique(lefse.out$direction))&lt;3){
g1&lt;-g1+scale_color_manual(values=c(Palette_LTBI_cured))+
scale_fill_manual(values=c(Palette_LTBI_cured))
}
print(g1)
</code></pre>

<p><img alt="lefse" src="../img/Cured/LEfSe_output.png" /></p>
<pre><code class="r">ltk&lt;-as.character(lefse.out$taxon)
phy_ra_ltk&lt;-prune_taxa(ltk,phy.lefse)
phy_ra_ltk_m&lt;-psmelt(phy_ra_ltk)

phy_ra_ltk_m$OTU&lt;-factor(phy_ra_ltk_m$OTU, levels=lefse.out$taxon[order(lefse.out$lda)])

g2&lt;-ggplot(phy_ra_ltk_m,aes(x=Genus,
y=Abundance,color=TB_status,
fill=TB_status))+
geom_boxplot(position=position_dodge(),
colour=&quot;black&quot;, # Use black outlines,
size=.3,alpha=0.5) +      # Thinner lines
theme_base()+  xlab(&quot;&quot;)+  coord_flip() +
scale_y_continuous(limits = c(0,300))
if(length(unique(lefse.out$direction))&lt;3){
g2&lt;-g2+scale_color_manual(values=c(Palette_LTBI_cured))+
scale_fill_manual(values=c(Palette_LTBI_cured))
}
print(g2)
</code></pre>

<p><img alt="barplot_cured" src="../img/Cured/LEfSe_barplot.png" /></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../Treatment_metagenomics/" class="btn btn-neutral float-right" title="Treatment">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../Treatment/" class="btn btn-neutral" title="Treatment"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../Treatment/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../Treatment_metagenomics/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js"></script>
      <script src="../search/require.js"></script>
      <script src="../search/search.js"></script>

</body>
</html>
